activityRetainedScope and fragmentScope are failing when activity is recreated due to configuration change.

**Describe the bug**
When you rotate your phone an Activity is recreated due to configuration change. If we use the `activityRetainedScope` and `fragmentScope` we will have a definition crash:

`Caused by: org.koin.core.error.NoBeanDefFoundException: No definition found for type 'com.example.myapplication4.Foo'. Check your Modules configuration and add missing type and/or qualifier!`

**To Reproduce**
Steps to reproduce the behavior:
1. 
```
class MainFragment : Fragment(), AndroidScopeComponent {
    override val scope: Scope by fragmentScope()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.main_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view.findViewById<TextView>(R.id.textView).text = get<Foo>()()
    }
}
```
2. 
```
class MainActivity : FragmentActivity(), AndroidScopeComponent {

    override val scope: Scope by activityRetainedScope()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_activity)
    }
}
```
3. 
```
class App : Application() {

    override fun onCreate() {
        super.onCreate()
        startKoin {
            androidLogger()
            androidContext(this@App)
            modules(
                module {
                    scope<MainActivity> {
                        scoped<Foo> { Foo() }
                    }
                }
            )
        }
    }
}
```
4. Rotate your device, split screen or popin it. Crash occurs.

**Expected behavior**
To not crash

**Koin module and version:**
 `koin-android:3.5.6`

**Reason**
1. When `activityRetainedActivity` (in Activity#A1) is used, we store newly created scope (Scope#Activity#A1) in `ScopeHandlerViewModel`.
2. Fragment calls `fragmentScope` which uses `Activity.getScopeId` to retrieve the correct id (Scope#Activity#A1) 
3. When Android recreates the Activity (Activity#2) the `activityRetainedActivity` retrieves stored scope (Scope#Activity#A1).
4. Fragment also gets recreated and calls the `fragmentScope` which insides tries to find Activity scope (Scope#Activity#A2) which is not present. Causing a dependency issue.

The implementation of `fragmentScope` retrieves parent (Activity) scope using following method:
```fun ComponentActivity.getScopeOrNull(): Scope? = getKoin().getScopeOrNull(getScopeId())```
where `getScopeId()` is ```fun <T : Any> T.getScopeId() = this::class.getFullName() + "@" + this.hashCode()```

When Android re-creates Activities (ie. spawns new Activity) due to configuration changes the `getScopeId` results in different values:
1. `MainActivity com.example.myapplication4.MainActivity@845bb25 - Foo com.example.myapplication4.Foo@c2f96c6`
2. Rotate screen
3. `MainActivity com.example.myapplication4.MainActivity@b67ace9 - Foo com.example.myapplication4.Foo@c2f96c6`

In my opinion fragmentScope shouldn't use the `this::class.getFullName() + "@" + this.hashCode()` to find Activity scope.
I was able to fix this by creating my own `fragmentScope` in following format:

```
fun Fragment.createFragmentScope(useParentActivityScope : Boolean = true): Scope {
    if (this !is AndroidScopeComponent) {
        error("Fragment should implement AndroidScopeComponent")
    }
    val scope = getKoin().getScopeOrNull(getScopeId()) ?: createScopeForCurrentLifecycle(this)
    if (useParentActivityScope){
        val activityScope = (requireActivity() as? AndroidScopeComponent)?.scope
        if (activityScope != null) {
            scope.linkTo(activityScope)
        } else {
            scope.logger.debug("Fragment '$this' can't be linked to parent activity scope")
        }
    }
    return scope
}
```

Which uses the ` val activityScope = (requireActivity() as? AndroidScopeComponent)?.scope` instead of `val activityScope = requireActivity().getScopeOrNull()`.

Please let me know if this is really an issue or my lack of understanding how scopes work in koin. 

Thanks :)

--- a/examples/androidx-samples/src/main/java/org/koin/sample/sandbox/di/AppModule.kt
+++ b/examples/androidx-samples/src/main/java/org/koin/sample/sandbox/di/AppModule.kt
@@ -28,6 +28,7 @@ import org.koin.sample.sandbox.mvvm.MVVMFragment
 import org.koin.sample.sandbox.navigation.NavViewModel
 import org.koin.sample.sandbox.navigation.NavViewModel2
 import org.koin.sample.sandbox.scope.ScopedActivityA
+import org.koin.sample.sandbox.scope.ScopedFragment
 import org.koin.sample.sandbox.workmanager.SimpleWorker
 import org.koin.sample.sandbox.workmanager.SimpleWorkerService
 
@@ -106,8 +107,12 @@ val scopeModule = lazyModule {
 
 val scopeModuleActivityA = lazyModule {
     scope<ScopedActivityA> {
+        fragmentOf(::ScopedFragment)
         scopedOf(::Session)
         scopedOf(::SessionActivity)
+    }
+    scope<ScopedFragment> {
+
     }
 }
 

--- a/examples/androidx-samples/src/main/java/org/koin/sample/sandbox/scope/ScopedActivityA.kt
+++ b/examples/androidx-samples/src/main/java/org/koin/sample/sandbox/scope/ScopedActivityA.kt
@@ -5,11 +5,13 @@ import android.widget.Button
 import org.koin.android.ext.android.get
 import org.koin.android.ext.android.getKoin
 import org.koin.android.ext.android.inject
+import org.koin.androidx.fragment.android.replace
 import org.koin.androidx.scope.RetainedScopeActivity
 import org.koin.core.qualifier.named
 import org.koin.sample.sandbox.R
 import org.koin.sample.sandbox.components.*
 import org.koin.sample.sandbox.components.scope.Session
+import org.koin.sample.sandbox.mvvm.MVVMFragment
 import org.koin.sample.sandbox.utils.navigateTo
 
 class ScopedActivityA : RetainedScopeActivity(R.layout.scoped_activity_a) {
@@ -19,6 +21,10 @@ class ScopedActivityA : RetainedScopeActivity(R.layout.scoped_activity_a) {
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+
+        supportFragmentManager.beginTransaction()
+            .replace<ScopedFragment>(R.id.mvvm_frame)
+            .commit()
 
         assert(currentSession == get<Session>())
         if (SESSION_ID_VAR.isEmpty()) {

--- a/examples/androidx-samples/src/main/java/org/koin/sample/sandbox/scope/ScopedFragment.kt
+++ b/examples/androidx-samples/src/main/java/org/koin/sample/sandbox/scope/ScopedFragment.kt
@@ -1,0 +1,28 @@
+package org.koin.sample.sandbox.scope
+
+import android.os.Bundle
+import android.view.View
+import androidx.fragment.app.Fragment
+import org.koin.android.scope.AndroidScopeComponent
+import org.koin.androidx.scope.fragmentScope
+import org.koin.androidx.scope.getRetainedScopeOrNull
+import org.koin.core.scope.Scope
+import org.koin.sample.sandbox.R
+import org.koin.sample.sandbox.components.scope.Session
+
+class ScopedFragment : Fragment(R.layout.mvvm_fragment), AndroidScopeComponent {
+
+    override val scope: Scope by fragmentScope()
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        checks()
+    }
+
+    private fun checks() {
+        assert(
+            scope.get<Session>() == requireActivity().getRetainedScopeOrNull()?.get<Session>()
+        )
+    }
+}

--- a/examples/androidx-samples/src/main/res/layout/scoped_activity_a.xml
+++ b/examples/androidx-samples/src/main/res/layout/scoped_activity_a.xml
@@ -18,4 +18,9 @@
         android:layout_height="wrap_content"
         android:text="Next" />
 
+    <FrameLayout
+        android:id="@+id/mvvm_frame"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"/>
+
 </LinearLayout>

--- a/projects/android/koin-android/src/main/java/org/koin/androidx/scope/ComponentActivityExt.kt
+++ b/projects/android/koin-android/src/main/java/org/koin/androidx/scope/ComponentActivityExt.kt
@@ -28,6 +28,7 @@ import org.koin.core.component.getScopeId
 import org.koin.core.component.getScopeName
 import org.koin.core.scope.Scope
 import org.koin.core.scope.ScopeCallback
+import org.koin.ext.getFullName
 
 /**
  * Provide Koin Scope tied to ComponentActivity
@@ -84,13 +85,14 @@ fun ComponentActivity.createActivityRetainedScope(): Scope {
     if (this !is AndroidScopeComponent) {
         error("Activity should implement AndroidScopeComponent")
     }
-//    if (this.scope != null) {
-//        error("Activity Scope is already created")
-//    }
+
     val scopeViewModel = viewModels<ScopeHandlerViewModel>().value
     if (scopeViewModel.scope == null) {
-        val scope = getKoin().createScope(getScopeId(), getScopeName())
+        val scope = getKoin().createScope(retainedScopeId(), getScopeName())
         scopeViewModel.scope = scope
     }
     return scopeViewModel.scope!!
 }
+
+fun ComponentActivity.retainedScopeId() : String = this::class.getFullName()
+fun ComponentActivity.getRetainedScopeOrNull(): Scope? = getKoin().getScopeOrNull(retainedScopeId())

--- a/projects/android/koin-android/src/main/java/org/koin/androidx/scope/FragmentExt.kt
+++ b/projects/android/koin-android/src/main/java/org/koin/androidx/scope/FragmentExt.kt
@@ -19,7 +19,6 @@ import androidx.fragment.app.Fragment
 import org.koin.android.ext.android.getKoin
 import org.koin.android.scope.AndroidScopeComponent
 import org.koin.core.component.getScopeId
-import org.koin.core.component.getScopeName
 import org.koin.core.scope.Scope
 
 /**
@@ -33,11 +32,11 @@ fun Fragment.createFragmentScope(useParentActivityScope : Boolean = true): Scope
     }
     val scope = getKoin().getScopeOrNull(getScopeId()) ?: createScopeForCurrentLifecycle(this)
     if (useParentActivityScope){
-        val activityScope = requireActivity().getScopeOrNull()
+        val activityScope = requireActivity().getScopeOrNull() ?: requireActivity().getRetainedScopeOrNull()
         if (activityScope != null) {
             scope.linkTo(activityScope)
         } else {
-            scope.logger.debug("Fragment '$this' can't be linked to parent activity scope")
+            scope.logger.debug("Fragment '$this' can't be linked to parent activity scope. No Parent Scope found.")
         }
     }
     return scope

